Follow-up fixes for the API rate limiter

tgraf opened the pull request:
So far, the rate-limiting has been enforced before enforcing parallel
requests. In theory, this is better because we can return error code
429 earlier to tell callers to slow down.

In practice, many callers will retry immediately anyway which means that
all the limiting will happen by the rate limiter. The rate limiter
relies on reservations that need to be canceled. If too many requests
happen in parallel, reservations can't be canceled quickly enough.

By swapping the enforcement of parallel requests with the rate limiter,
all requests will block for at least MaxWaitDuration if more than the
allowed number of parallel requests are pending which will naturally
pace the callers.

As the semaphore enforcing parallel requests now has to be released in
error scenarios, change the structure of Wait() to have a single error
handling structure. By reusing finishRequest(), the metrics handler has
to be adjusted slightly to account for new outcomes as it now bumps the
metric for canceled requests as well.

tgraf committed:
api-limiter: Add test covering cancelation of rate limiter

Test that failing requests will not impact the rate limiter. This is
already working correctly so this test only adds additional coverage.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
api-limiter: Fix duplicate log message

The following log message was duplicated and printed twice for two
different meanings.

Fixes: 3141e6581e0 ("rate: Add API rate limiting system")
Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
api-limiter: Enforce ParallelRequests before applying rate limiting

So far, the rate limiting has been enforced before enforcing parallel
requests. In theory, this is better because we can return error code
429 earlier to tell callers to slow down.

In practice, many callers will retry immediately anyway which means that
all the limiting will happen by the rate limiter. The rate limiter
relies on reservations that need to be canceled. If too many requests
happen in parallel, reservations can't be canceled quickly enough.

By swapping the enforcement of parallel requests with the rate limiter,
all requests will block for at least MaxWaitDuration if more than the
allowed number of parallel requests are pending which will naturally
pace the callers.

Swapping the enforcement order requires the acquired semaphore to be
released in error cases of the rate limiter. This requires to change the
structure of Wait() to have a single error handling structure. By
reusing finishRequest(), the metrics handler has to be adjusted slightly
to account for new outcomes as it now bumps the metric for canceled
requests as well. What remains unchanged is that only successful API
requests are used to calculate the mean processing duration.

Fixes: 3141e6581e0 ("rate: Add API rate limiting system")
Signed-off-by: Thomas Graf <thomas@cilium.io>

christarazi commented on pkg/rate/api_limiter.go:
```diff
@@ -409,13 +416,18 @@ func (l *APILimiter) adjustedParallelRequests() int {
 }
 
 func (l *APILimiter) requestFinished(r *limitedRequest, err error) {
+	var processingDuration time.Duration
```
Style nit: we can move this right above the `if` block before it's used.

joestringer commented on pkg/rate/api_limiter_test.go:
```diff
@@ -716,3 +718,78 @@ func (b *ControllerSuite) TestAdjustedParallelRequests(c *check.C) {
+	var completed int32
+
+	// All of these requests must fail due to having to wait too long as
+	// the only parallel request slot is occupied. The rate limiter should
+	// not get occupied with these requests though.
```
> The rate limiter should not get occupied with these requests though.

Did you intend to test this as well?

This test passes on master, regardless of the patches. Even the error returned by each goroutine is the same with/without this PR:

```
timed out while waiting to be served with 1 parallel requests: context deadline exceeded
```

The test itself looks like a reasonable assessment of the behaviour though, just checking whether it was intended as a general improvement in coverage or whether you were aiming to test the functionality of this PR specifically.

tgraf replied:
I've added the test to validate the cancelation of the parallel request and did not find any bugs in this regard in the existing code. I figured it makes sense to have the test in the codebase though to catch eventual regressions.

joestringer commented on pkg/rate/api_limiter.go:
```diff
@@ -618,16 +630,69 @@ func (l *APILimiter) Wait(ctx context.Context) (req LimitedRequest, err error) {
+
+	l.mutex.Lock()
+	if l.limiter != nil {
+		r = l.limiter.Reserve()
```
:thinking: Is there a way to make the corresponding `r.Cancel()` more obvious? The previous deferral made it trivial to follow and balance (and potentially modify this code in future without unintentionally leaking it).

tgraf replied:
Yes, unfortunately, the defer did not longer work because the `time.Sleep()` should happen after the `r.Cancel()`.

joestringer commented on pkg/rate/api_limiter.go:
```diff
@@ -442,30 +454,34 @@ func (l *APILimiter) requestFinished(r *limitedRequest, err error) {
-		l.waitDurations = l.waitDurations[exceed:]
-	}
-	l.meanWaitDuration = calcMeanDuration(l.waitDurations)
+	if err == nil {
```
nit, this context might be useful

```suggestion
	// Only auto-adjust ratelimiter using metrics from successful API requests
	if err == nil {
```

joestringer reviewed (approved):
The change looks correct to me, though I'm not sure whether the tests are validating the change in functionality.

christarazi reviewed (commented):
The first 2 commits are fairly straightforward so they look good, but the 3rd commit required closer inspection. 

I was able to identity 3 major changes in 3rd commit:

  1)  Only updating the wait duration / mean duration times when a call succeeds
   2)  Cancelling reservations if it exceeds the max wait duration time
   3) Sleeping in order to artificially pace against stubborn clients 

(1) makes sense. (2) makes sense at a high-level, but I was expecting the test for it to fail with the prior rate limiting implementation. I'd also like to see a test for (3).

FWIW, the stress test also passes with the prior rate limiting implementation. This makes it difficult to assess the code changes in this PR.

tgraf committed:
api-limiter: Add stress test

Simulate stress by feeding many parallel requests through the API rate
limiter. The maximum wait duration will soon be hit for requests causing
them to fail.

Without the fix, this test would often fail. If it succeeded, the number
of retries is in the range of 60-80K. With the fix, the test succeeds
consistently and retries are in the range of 6-8K. That's a 10x
reduction of retries.

Signed-off-by: Thomas Graf <thomas@cilium.io>

PinchasLev commented on pkg/rate/api_limiter_test.go:
```diff
@@ -716,3 +718,91 @@ func (b *ControllerSuite) TestAdjustedParallelRequests(c *check.C) {
+
+	var completed int32
+
+	// All of these requests must fail due to having to wait too long as
```
Looking to clarify, whether these 20 requests fail not because the 500ms wait time is exceeded, but because the one and only parallel slot is taken, and when the slot is taken an attempted request fails immediately.  That's not what I expect.

It is my expectation that the original request will also time out and fail prior to req.Done() being called.  Though that doesn't alter the test, in the sense that the final 10 requests are still expected to succeed, given that the prior 21 (20 in loop and the 1 original) requests have failed.

PinchasLev replied:
@tgraf Moreover, I'd like to add, that my understanding is that at some point the original request would timeout, and one of the 20 requests in the loop would replace it in the single parallel slot.  Only, to subsequently be cancelled itself for never completing.  I say so, because the original request should time out before the 20 that came after it.

tgraf replied:
Yes, the 20 requests will fail because the only parallel slot available is still held by the request created before the loop. All 20 requests will sit there, waiting for 500ms, and then timeout. The test verifies that the reservation is properly canceled so the next request will immediately succeed.

> It is my expectation that the original request will also time out and fail prior to req.Done() being called.

The first request will succeed as it is already being processed. By the time `Wait()` returns for the first request, the rate limiter has given the OK for the API request to be processed. While it is being processed, the 20 requests get queued up, time out and the the initial request signals completion by calling `req.Done()`. The `MaxWaitDuration` only applies to the waiting until the API request is allowed to be processed.

PinchasLev replied:
My bad.  I overlooked that the it's "waitDuration", not your traditional request timeout value.

christarazi commented on pkg/rate/api_limiter_test.go:
```diff
@@ -716,3 +718,49 @@ func (b *ControllerSuite) TestAdjustedParallelRequests(c *check.C) {
+		return atomic.LoadInt32(&completed) == 1000
+	}, 5*time.Second), check.IsNil)
+
+	log.Infof("%+v", a)
+	log.Infof("Total retries: %v", atomic.LoadInt32(&retries))
```
We can convert these into `c.Log` instead, so we don't pollute the output on successful runs. I imagine this was only useful for debugging.

christarazi reviewed (approved):
Nice! I am able to confirm that the stress test fails, but I wasn't for the cancel one. I think that's fine. Non-blocking nit that doesn't need a full CI run.
